Add recipe for flymake-rest

mohkale opened the pull request:
Note: This continues from #7634.

### Brief summary of what the package does

Provide helpers for replacing flycheck with flymake including a flycheck like error parser and several builtin checker definitions.

### Direct link to the package repository

https://github.com/mohkale/flymake-rest

### Your association with the package

Package Author.

### Relevant communications with the upstream package maintainer

None needed.

### Checklist

<!-- Please confirm with `x`: -->

**Note**: I plan to complete these soon.

- [X] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [X] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [X] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [X] My elisp byte-compiles cleanly
- [ ] `M-x checkdoc` is happy with my docstrings
- [ ] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

<!-- After submitting, please fix any problems the CI reports. -->

riscy commented:
Thanks for this!  I notice you haven't checked the `checkdoc` box.  I (personally) usually ask that contributors do their best with these, even though they might disagree with them, because frankly it reduces the amount of thinking I have to do to rationalize the grey areas. :)

In the meantime, here's a quick first pass over those checklist items -- including some package-lints you may not have caught:

`flymake-rest-awk-gawk.el` with checkdoc 0.6.2:
```
flymake-rest-awk-gawk.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-clang.el` with checkdoc 0.6.2:
```
flymake-rest-clang.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-commands.el` with checkdoc 0.6.2:
```
flymake-rest-commands.el:0: You should have a section marked ";;; Commentary:"
flymake-rest-commands.el:29: All variables and subroutines might as well have a documentation string
flymake-rest-commands.el:50: All variables and subroutines might as well have a documentation string
flymake-rest-commands.el:76: Argument `checkers' should appear (as CHECKERS) in the doc string
```

`flymake-rest-define.el` with checkdoc 0.6.2:
```
flymake-rest-define.el:168: Lisp symbol `standard-input' should appear in quotes
flymake-rest-define.el:462: Argument `name' should appear (as NAME) in the doc string
```

`flymake-rest-define.el` with package-lint 20211122.1152:
```
1 issue found:
50:0: warning: Private functions generally should not be autoloaded.
```

`flymake-rest-eslint.el` with checkdoc 0.6.2:
```
flymake-rest-eslint.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-gcc.el` with checkdoc 0.6.2:
```
flymake-rest-gcc.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-hook.el` with checkdoc 0.6.2:
```
flymake-rest-hook.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-hook.el` with package-lint 20211122.1152:
```
5 issues found:
130:1: warning: `with-eval-after-load' is for use in configurations, and should rarely be used in packages.
144:0: error: "use-package-normalize/:flymake-hook" doesn't start with package's prefix "flymake-rest".
144:0: error: `use-package-normalize/:flymake-hook' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
147:0: error: "use-package-handler/:flymake-hook" doesn't start with package's prefix "flymake-rest".
147:0: error: `use-package-handler/:flymake-hook' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
```

`flymake-rest-html-tidy.el` with checkdoc 0.6.2:
```
flymake-rest-html-tidy.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-jq.el` with checkdoc 0.6.2:
```
flymake-rest-jq.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-jsonlint.el` with checkdoc 0.6.2:
```
flymake-rest-jsonlint.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-less.el` with checkdoc 0.6.2:
```
flymake-rest-less.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-lua.el` with checkdoc 0.6.2:
```
flymake-rest-lua.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-luacheck.el` with checkdoc 0.6.2:
```
flymake-rest-luacheck.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-luacheck.el` with package-lint 20211122.1152:
```
1 issue found:
42:0: error: "flymake-luacheck-rc" doesn't start with package's prefix "flymake-rest".
```

`flymake-rest-markdownlint.el` with checkdoc 0.6.2:
```
flymake-rest-markdownlint.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-mypy.el` with checkdoc 0.6.2:
```
flymake-rest-mypy.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-proselint.el` with checkdoc 0.6.2:
```
flymake-rest-proselint.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-pycodestyle.el` with checkdoc 0.6.2:
```
flymake-rest-pycodestyle.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-pylint.el` with checkdoc 0.6.2:
```
flymake-rest-pylint.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-rubocop.el` with checkdoc 0.6.2:
```
flymake-rest-rubocop.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-shellcheck.el` with checkdoc 0.6.2:
```
flymake-rest-shellcheck.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-sql-lint.el` with checkdoc 0.6.2:
```
flymake-rest-sql-lint.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-sqlint.el` with checkdoc 0.6.2:
```
flymake-rest-sqlint.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-xmllint.el` with checkdoc 0.6.2:
```
flymake-rest-xmllint.el:0: You should have a section marked ";;; Commentary:"
```

`flymake-rest-yamllint.el` with checkdoc 0.6.2:
```
flymake-rest-yamllint.el:0: You should have a section marked ";;; Commentary:"
```

mohkale commented:
@riscy 

I don't really see the value in having a commentary in each checker file, but I've done so now. I've also added a makefile to run automate running these checks and fixed all the issues raised by it. Please let me know if anything else needs to be done :-).

riscy commented on recipes/flymake-rest:
```diff
@@ -0,0 +1,4 @@
+(flymake-rest
+ :fetcher github
+ :repo "mohkale/flymake-rest"
+ :files ("*.el" "checkers/*.el"))
```
```suggestion
 :files (:defaults "checkers/*.el"))
```

riscy reviewed (commented):
Thanks for accommodating!  You are right it is probably needlessly pedantic, but it's also something we ask of all packages across the board to be fair to everyone and (automatically) maintain some minimal standards.

I notice your recipe is worth changing to include `:defaults` which will automatically ignore certain files prefixed with `test-` and so on -- allowing your package to grow without opening another PR.  Suggestion included.

And I ran another check and I think there is a missing quotation mark in the package-requires of flymake-rest.el that package-lint will catch for you.

mohkale commented:
> And I ran another check and I think there is a missing quotation mark in the package-requires of flymake-rest.el that package-lint will catch for you.

mohkale commented:
> And I ran another check and I think there is a missing quotation mark in the package-requires of flymake-rest.el that package-lint will catch for you.

Wow, thanks for spotting that. I'll add another makefile entry to run package-lint as well. I avoided it before since it complains about missing package-requires declarations in files beside the main `flymake-rest.el` but I should have it for that file at least. :-).

riscy commented:
Thanks, that was getting my tool stuck but it's all fixed now.  Here's all that's left:

`flymake-rest-define.el` with package-lint 20211122.1152:
```
1 issue found:
50:0: warning: Private functions generally should not be autoloaded.
```

`flymake-rest-hook.el` with package-lint 20211122.1152:
```
5 issues found:
132:1: warning: `with-eval-after-load' is for use in configurations, and should rarely be used in packages.
146:0: error: "use-package-normalize/:flymake-hook" doesn't start with package's prefix "flymake-rest".
146:0: error: `use-package-normalize/:flymake-hook' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
149:0: error: "use-package-handler/:flymake-hook" doesn't start with package's prefix "flymake-rest".
149:0: error: `use-package-handler/:flymake-hook' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
```
We try to be pretty strict about prefixes whenever it's possible, because it simplifies resolving namespace conflicts.

mohkale commented:
> 50:0: warning: Private functions generally should not be autoloaded.

That's actually a private variable but autoloading it prevents having to load `flymake-rest-define` after it's been compiled. I could try making it non-private but I think the current approach is best.

> 132:1: warning: `with-eval-after-load' is for use in configurations, and should rarely be used in packages.
> 146:0: error: "use-package-normalize/:flymake-hook" doesn't start with package's prefix "flymake-rest".
> 146:0: error: `use-package-normalize/:flymake-hook' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
> 149:0: error: "use-package-handler/:flymake-hook" doesn't start with package's prefix "flymake-rest".
> 149:0: error: `use-package-handler/:flymake-hook' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).

All of that is related to use-package and I believe use-package has strict requirements on what form it's keyword handlers should have so I can't change it.

riscy commented:
I read up on `use-package` and I agree.  I'll merge -- thanks!

mohkale commented:
Much appreciated, glad to have this merged :-).
